Lua getDateTime() returns integer values. I am suggesting an enhancement.

When we use the getDateTime() routine in lua scripts, the table that is created is integer values. The means that if we are putting together a string like YYYYMMDD, there are no leading zeros on values below 10. January 1, 2015 would come together as 201511. The question is, do we want to fix this using lua, which I have a routine written for, or do we want to fix this in the opentx model code. I have written and tested the following code in radio/src/lua/api_general.cpp. 

```c
static void luaPushDateTimeStr(lua_State *L, uint32_t year, uint32_t mon, uint32_t day,
                            uint32_t hour, uint32_t min, uint32_t sec)
{
  char stryear[5];
  char strmon[3];
  char strday[3];
  char strhour[3];
  char strmin[3];
  char strsec[3];
  sprintf(stryear, "%lu", year);
  if (mon<10) sprintf(strmon, "0%lu", mon); else 	sprintf(strmon, "%lu", mon);
  if (day<10) sprintf(strday, "0%lu", day); else sprintf(strday, "%lu", day);
  if (hour<10) sprintf(strhour, "0%lu", hour); else sprintf(strhour, "%lu", hour);
  if (min<10) sprintf(strmin, "0%lu", min); else sprintf(strmin, "%lu", min);
  if (sec<10) sprintf(strsec, "0%lu", sec); else sprintf(strsec, "%lu", sec);
  lua_newtable(L);
  lua_pushtablestring(L, "year", stryear);
  lua_pushtablestring(L, "mon", strmon);
  lua_pushtablestring(L, "day", strday);
  lua_pushtablestring(L, "hour", strhour);
  lua_pushtablestring(L, "min", strmin);
  lua_pushtablestring(L, "sec", strsec);
}

/*luadoc
@function getDateTimeStr()

Returns current system date and time that is kept by the RTC unit as a string with leading zeros

@retval table current date and time, table elements:
 * `year` year
 * `mon` month
 * `day` day of month
 * `hour` hours
 * `min` minutes
 * `sec` seconds
*/

static int luaGetDateTimeStr(lua_State *L)
{
  struct gtm utm;
  gettime(&utm);
  luaPushDateTimeStr(L, utm.tm_year + 1900, utm.tm_mon + 1, utm.tm_mday, utm.tm_hour, utm.tm_min, utm.tm_sec);
  return 1;
}
```

As a secondary issue, I am compiling this on a 32 bit system:
```
isaac@isaac-ThinkPad-T61:~/myopentx/opentx/companion/lbuild$ uname -m
i686
isaac@isaac-ThinkPad-T61:~/myopentx/opentx/companion/lbuild$ getconf LONG_BIT
32
```

If I use %lu in the sprintf command above, I get a warning when I compile Companion:
```
/home/isaac/myopentx/opentx/companion/src/../../radio/src/lua/api_general.cpp:168:31: warning: format ‘%lu’ expects argument of type ‘long unsigned int’, but argument 3 has type ‘uint32_t {aka unsigned int}’ [-Wformat]
```
and no warning when I compile the Model.

If I change it to %u, I get no warning when compiling Companion and this warning when compiling Model (make PCB=TARANIS LUA=YES):
```
lua/api_general.cpp:168:30: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'uint32_t {aka long unsigned int}' [-Wformat]
```

Looks like there are two issues here.
1) Do we want to solve the leading zero problem on the getDateTime with a new getDateTimeStr function in the radio code, or would we rather keep the solution on the lua side. I am honestly not sure which would save more space, but know that we want to keep the solution in the safest most efficient place. If we decide to go with this, I will request a pull with my new code after completing my testing. It's working in the simulator but I want to test it in my radio. 

2) What is the proper sprintf conversion for uint32_t? I can only find %u and %lu, and either one or the other compiles throws a warning.

--- a/companion/src/firmwares/opentx/simulator/opentxsimulator.cpp
+++ b/companion/src/firmwares/opentx/simulator/opentxsimulator.cpp
@@ -316,6 +316,7 @@ namespace NAMESPACE {
 #include "radio/src/thirdparty/Lua/src/ltablib.c"
 #include "radio/src/thirdparty/Lua/src/lcorolib.c"
 #include "radio/src/thirdparty/Lua/src/liolib.c"
+#include "radio/src/thirdparty/Lua/src/lstrlib.c"
 #endif
 
 int16_t g_anas[NUM_STICKS+NUM_POTS];

--- a/radio/src/Makefile
+++ b/radio/src/Makefile
@@ -399,7 +399,12 @@ IMG2LBM = -$(PYTHON) ../util/img2lbm.py
 BIN2LBM = -$(PYTHON) ../util/bin2lbm.py
 FONT2PNG = -$(PYTHON) ../util/font2png.py
 TRANSLATE = -$(PYTHON) ../util/translate.py
-VERSION = 2.1.6$(OPENTX_VERSION_SUFFIX)
+
+VERSION_MAJOR = 2
+VERSION_MINOR = 1
+VERSION_REVISION  = 6
+
+VERSION = $(VERSION_MAJOR).$(VERSION_MINOR).$(VERSION_REVISION)$(OPENTX_VERSION_SUFFIX)
 
 THIRDPARTY = thirdparty
 COOSDIR = $(THIRDPARTY)/CoOS
@@ -1538,6 +1543,10 @@ stamp_header:
 	@echo "#define DATE_STR \"`date +%Y-%m-%d | tr -d '\r\n'`\"" >> stamp-opentx.h
 	@echo "#define TIME_STR \"`date +%H:%M:%S | tr -d '\r\n'`\"" >> stamp-opentx.h
 	@echo "#define VERS_STR \"$(VERSION)\"" >> stamp-opentx.h
+	@echo "" >> stamp-opentx.h
+	@echo "#define VERSION_MAJOR     $(VERSION_MAJOR)" >> stamp-opentx.h
+	@echo "#define VERSION_MINOR     $(VERSION_MINOR)" >> stamp-opentx.h
+	@echo "#define VERSION_REVISION  $(VERSION_REVISION)" >> stamp-opentx.h
 
 opentx-stamp:
 	@echo "#define DATE_STR \"`date +%Y-%m-%d`\"" > stamp-opentx.txt

--- a/radio/src/Makefile
+++ b/radio/src/Makefile
@@ -1547,11 +1547,13 @@ stamp_header:
 	@echo "#define VERSION_MAJOR     $(VERSION_MAJOR)" >> stamp-opentx.h
 	@echo "#define VERSION_MINOR     $(VERSION_MINOR)" >> stamp-opentx.h
 	@echo "#define VERSION_REVISION  $(VERSION_REVISION)" >> stamp-opentx.h
+	@echo "#define GIT_STR \"$(shell git rev-parse HEAD)\"" >> stamp-opentx.h
 
 opentx-stamp:
 	@echo "#define DATE_STR \"`date +%Y-%m-%d`\"" > stamp-opentx.txt
 	@echo "#define TIME_STR \"`date +%H:%M:%S`\"" >> stamp-opentx.txt
 	@echo "#define VERS_STR \"$(VERSION)\"" >> stamp-opentx.txt
+	@echo "#define GIT_STR \"$(shell git rev-parse HEAD)\"" >> stamp-opentx.txt
 	@cat stamp-opentx.txt
 
 translations/en.h: translations/en.h.txt

--- a/radio/src/Makefile
+++ b/radio/src/Makefile
@@ -1036,7 +1036,7 @@ ifeq ($(PCB), TARANIS)
     CPPSRC += lua/interface.cpp lua/api_general.cpp lua/api_lcd.cpp lua/api_model.cpp
     LUASRC = $(LUADIR)/lapi.c $(LUADIR)/lcode.c $(LUADIR)/lctype.c $(LUADIR)/ldebug.c $(LUADIR)/ldo.c $(LUADIR)/ldump.c $(LUADIR)/lfunc.c $(LUADIR)/lgc.c $(LUADIR)/llex.c $(LUADIR)/lmem.c \
            $(LUADIR)/lobject.c $(LUADIR)/lopcodes.c $(LUADIR)/lparser.c $(LUADIR)/lstate.c $(LUADIR)/lstring.c $(LUADIR)/ltable.c $(LUADIR)/lrotable.c $(LUADIR)/ltm.c $(LUADIR)/lundump.c $(LUADIR)/lvm.c $(LUADIR)/lzio.c \
-           $(LUADIR)/lbaselib.c $(LUADIR)/linit.c $(LUADIR)/lmathlib.c $(LUADIR)/lbitlib.c $(LUADIR)/loadlib.c $(LUADIR)/lauxlib.c $(LUADIR)/ltablib.c $(LUADIR)/lcorolib.c $(LUADIR)/liolib.c
+           $(LUADIR)/lbaselib.c $(LUADIR)/linit.c $(LUADIR)/lmathlib.c $(LUADIR)/lbitlib.c $(LUADIR)/loadlib.c $(LUADIR)/lauxlib.c $(LUADIR)/ltablib.c $(LUADIR)/lcorolib.c $(LUADIR)/liolib.c $(LUADIR)/lstrlib.c
     SRC += $(LUASRC)
     LUADEP = lua_exports.inc
 	ifeq ($(USE_BIN_ALLOCATOR), YES)

--- a/radio/src/lua/api_general.cpp
+++ b/radio/src/lua/api_general.cpp
@@ -72,22 +72,28 @@ Return OpenTX version
 
 @retval string OpenTX version (ie "2.1.5")
 
-@retval list (available since 2.1.7) returns two values:
- * `string` OpenTX version (ie "2.1.5")
- * `string` radio version: `taranisx9e`, `taranisplus` or `taranis`. 
+@retval multiple (available since 2.1.7) returns 5 values:
+ * (string) OpenTX version (ie "2.1.5")
+ * (string) radio version: `taranisx9e`, `taranisplus` or `taranis`. 
 If running in simulator the "-simu" is added
+ * (number) major version (ie 2 if version 2.1.5)
+ * (number) minor version (ie 1 if version 2.1.5)
+ * (number) revison number (ie 5 if version 2.1.5)
 
 @status current Introduced in 2.0.0, expanded in 2.1.7
 
 ### Example
 
-This example also runs in OpenTX versions where the radio version was not available:
+This example also runs in OpenTX versions where the function returned only one value:
 
 ```lua
 local function run(event)
-  local ver, radio = getVersion()
+  local ver, radio, maj, minor, rev = getVersion()
   print("version: "..ver)
   if radio then print ("radio: "..radio) end
+  if maj then print ("maj: "..maj) end
+  if minor then print ("minor: "..minor) end
+  if rev then print ("rev: "..rev) end
   return 1
 end
 
@@ -97,14 +103,19 @@ Output of the above script in simulator:
 ```
 version: 2.1.7
 radio: taranis-simu
-Script finished with status 1
+maj: 2
+minor: 1
+rev: 7
 ```
 */
 static int luaGetVersion(lua_State *L)
 {
   lua_pushstring(L, VERS_STR);
   lua_pushstring(L, RADIO_VERSION);
-  return 2;
+  lua_pushnumber(L, VERSION_MAJOR);
+  lua_pushnumber(L, VERSION_MINOR);
+  lua_pushnumber(L, VERSION_REVISION);
+  return 5;
 }
 
 /*luadoc

--- a/radio/src/lua/api_general.cpp
+++ b/radio/src/lua/api_general.cpp
@@ -170,20 +170,15 @@ static int luaGetDateTime(lua_State *L)
 static void luaPushLatLon(TelemetrySensor & telemetrySensor, TelemetryItem & telemetryItem)
 /* result is lua table containing members ["lat"] and ["lon"] as lua_Number (doubles) in decimal degrees */
 {
-  lua_Number lat = 0.0;
-  lua_Number lon = 0.0;
   uint32_t gpsLat = 0;
   uint32_t gpsLon = 0;
-
   telemetryItem.gps.extractLatitudeLongitude(&gpsLat, &gpsLon); /* close, but not the format we want */
-  lat = gpsLat / 1000000.0;
-  if (telemetryItem.gps.latitudeNS == 'S') lat = -lat;
-  lon = gpsLon / 1000000.0;
-  if (telemetryItem.gps.longitudeEW == 'W') lon = -lon;
 
-  lua_createtable(L, 0, 2);
-  lua_pushtablenumber(L, "lat", lat);
-  lua_pushtablenumber(L, "lon", lon);
+  lua_createtable(L, 0, 4);
+  lua_pushtablenumber(L, "lat", gpsLat / ((telemetryItem.gps.latitudeNS == 'S') ? -1000000.0 : 1000000.0));
+  lua_pushtablenumber(L, "pilot-lat", telemetryItem.pilotLatitude / ((telemetryItem.gps.latitudeNS == 'S') ? -1000000.0 : 1000000.0));
+  lua_pushtablenumber(L, "lon", gpsLon / ((telemetryItem.gps.longitudeEW == 'W') ? -1000000.0 : 1000000.0));
+  lua_pushtablenumber(L, "pilot-lon", telemetryItem.pilotLongitude / ((telemetryItem.gps.longitudeEW == 'W') ? -1000000.0 : 1000000.0));
 }
 
 static void luaPushTelemetryDateTime(TelemetrySensor & telemetrySensor, TelemetryItem & telemetryItem)
@@ -381,6 +376,8 @@ or a name (string) of the source.
 @retval table GPS position is returned in a table:
  * `lat` (number) latitude, positive is North 
  * `lon` (number) longitude, positive is East
+ * `pilot-lat` (number) pilot latitude, positive is North 
+ * `pilot-lon` (number) pilot longitude, positive is East
 
 @retval table GPS date/time, see getDateTime()
 

--- a/radio/src/targets/simu/simpgmspace.cpp
+++ b/radio/src/targets/simu/simpgmspace.cpp
@@ -822,7 +822,7 @@ FRESULT f_open (FIL * fil, const TCHAR *name, BYTE flag)
   fil->fs = (FATFS*)fopen(realPath, (flag & FA_WRITE) ? ((flag & FA_CREATE_ALWAYS) ? "wb+" : "ab+") : "rb+");
   fil->fptr = 0;
   if (fil->fs) {
-    TRACE("f_open(%s, %x) = %p", path, flag, (FILE*)fil->fs);
+    TRACE("f_open(%s, %x) = %p (FIL %p)", path, flag, fil->fs, fil);
     return FR_OK;
   }
   TRACE("f_open(%s) = error %d (%s)", path, errno, strerror(errno));
@@ -871,8 +871,9 @@ UINT f_size(FIL* fil)
 
 FRESULT f_close (FIL * fil)
 {
-  if (fil && fil->fs) {
-    TRACE("f_close(%p)", (FILE*)fil->fs);
+  assert(fil);
+  TRACE("f_close(%p) (FIL:%p)", fil->fs, fil);
+  if (fil->fs) {
     fclose((FILE*)fil->fs);
     fil->fs = NULL;
   }

--- a/radio/src/thirdparty/Lua/src/lauxlib.h
+++ b/radio/src/thirdparty/Lua/src/lauxlib.h
@@ -28,6 +28,7 @@ extern const luaL_Reg baselib[];
 extern const luaL_Reg mathlib[];
 extern const luaL_Reg bitlib[];
 extern const luaL_Reg iolib[];
+extern const luaL_Reg strlib[];
 
 extern const luaL_Reg opentxLib[];
 extern const luaL_Reg lcdLib[];

--- a/radio/src/thirdparty/Lua/src/linit.c
+++ b/radio/src/thirdparty/Lua/src/linit.c
@@ -46,6 +46,7 @@ const luaR_table lua_rotable[] =
   {LUA_MATHLIBNAME, mathlib, mathlib_vals},
   {LUA_BITLIBNAME, bitlib, NULL},
   {LUA_IOLIBNAME, iolib, NULL},
+  {LUA_STRLIBNAME, strlib, NULL},
   {"lcd", lcdLib, NULL},
   {"model", modelLib, NULL},
   {"__baselib", baselib, NULL},

--- a/radio/src/thirdparty/Lua/src/liolib.c
+++ b/radio/src/thirdparty/Lua/src/liolib.c
@@ -203,16 +203,17 @@ static int io_close (lua_State *L) {
 #endif
 }
 
-#if !defined(USE_FATFS)
 static int f_gc (lua_State *L) {
 #if !defined(USE_FATFS)
   LStream *p = tolstream(L);
   if (!isclosed(p) && p->f != NULL)
     aux_close(L);  /* ignore closed and incompletely open files */
 #endif
+  f_close(tofile(L));  // no need to check if file was already closed (fatfs will not close it if p->f->fs is 0)
   return 0;
 }
 
+#if !defined(USE_FATFS)
 /*
 ** function to close regular files
 */
@@ -659,31 +660,27 @@ const luaL_Reg iolib[] = {
   {NULL, NULL}
 };
 
-#if !defined(USE_FATFS)
-
 /*
 ** methods for file handles
 */
 static const luaL_Reg flib[] = {
-  {"close", io_close},
-  {"flush", f_flush},
-  {"lines", f_lines},
-  {"read", io_read},
-  {"seek", f_seek},
-  {"setvbuf", f_setvbuf},
-  {"write", f_write},
-  {"__gc", f_gc},
-  {"__tostring", f_tostring},
+  // {"close", io_close},
+  // {"flush", f_flush},
+  // {"lines", f_lines},
+  // {"read", io_read},
+  // {"seek", f_seek},
+  // {"setvbuf", f_setvbuf},
+  // {"write", f_write},
+  {"__gc", f_gc},     // we define only garbage collector to close any leftover open files
+  // {"__tostring", f_tostring},
   {NULL, NULL}
 };
-
-#endif
 
 static void createmeta (lua_State *L) {
   luaL_newmetatable(L, LUA_FILEHANDLE);  /* create metatable for file handles */
   lua_pushvalue(L, -1);  /* push metatable */
-  // lua_setfield(L, -2, "__index");  /* metatable.__index = metatable */
-  // luaL_setfuncs(L, flib, 0);  /* add file methods to new metatable */
+  lua_setfield(L, -2, "__index");  /* metatable.__index = metatable */
+  luaL_setfuncs(L, flib, 0);  /* add file methods to new metatable */
   lua_pop(L, 1);  /* pop new metatable */
 }
 

--- a/radio/src/thirdparty/Lua/src/lstrlib.c
+++ b/radio/src/thirdparty/Lua/src/lstrlib.c
@@ -9,6 +9,7 @@
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
+#include <stdio.h>
 
 #define lstrlib_c
 
@@ -975,7 +976,7 @@ static int str_format (lua_State *L) {
 /* }====================================================== */
 
 
-static const luaL_Reg strlib[] = {
+const luaL_Reg strlib[] = {
   {"byte", str_byte},
   {"char", str_char},
   {"dump", str_dump},
@@ -993,7 +994,7 @@ static const luaL_Reg strlib[] = {
   {NULL, NULL}
 };
 
-
+#if 0
 static void createmetatable (lua_State *L) {
   lua_createtable(L, 0, 1);  /* table to be metatable for strings */
   lua_pushliteral(L, "");  /* dummy string */
@@ -1014,4 +1015,4 @@ LUAMOD_API int luaopen_string (lua_State *L) {
   createmetatable(L);
   return 1;
 }
-
+#endif

